### PR TITLE
refactor representations as module

### PR DESCRIPTION
**Pull request by nstarman**
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

It's a long file that does many things. Now it's one module with many files, each of which do fewer things.

~For reviewers: I didn't change the test files, only moved them, so that should show that no (tested) functionality has changed!~

TODO?
- [x] changelog, is one needed for an internal refactor?
- [x] break up the test files? Low priority, perhaps push to future PR.
- [x] ~BREAKING CHANGE: change folder name to ``representations``~

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by nstarman** [comment]
@astropy/testing-infrastructure-maintainers  ``pytest_doctestplus`` is broken?

**Comment by mhvk** [comment]
Sorry pressed the "convert to real PR" in the wrong PR. Though I like the idea, so I guess we might as well.

File: astropy/coordinates/representation/__init__.py
Comment (mhvk) [inline astropy/coordinates/representation/__init__.py:17]: Is the idea to import these only for backward compatibility? If so, probably good to comment this explicitly.

File: astropy/coordinates/representation/cartesian.py
Comment (mhvk) [inline astropy/coordinates/representation/cartesian.py:19]: No need for this separation, I think

File: astropy/coordinates/representation/cylindrical.py
Comment (mhvk) [inline astropy/coordinates/representation/cylindrical.py:17]: Why is `__all__` empty? I think it can actually be omitted here, but if included, I'd just do the two classes; one can ensure that the module is not included in the docs (which would make sense).

(Also a bit odd to add typing information here only)

File: astropy/coordinates/representation/physicsphere.py
Comment (mhvk) [inline astropy/coordinates/representation/physicsphere.py:20]: Same two comments about `__all__` and `# CODE`

File: astropy/coordinates/representation/tests/test_representation.py
Comment (mhvk) [inline astropy/coordinates/representation/tests/test_representation.py:2]: Not sure we should move the tests - we don't for things in `units`.

File: astropy/coordinates/representation/tests/test_representation_arithmetic.py
Comment (mhvk) [inline astropy/coordinates/representation/tests/test_representation_arithmetic.py:3]: But if we move tests, we should remove `representation` from the test file name. So, `test_basics` above, `test_arithmetic` here.

File: astropy/utils/compat/numpycompat.py
Comment (mhvk) [inline astropy/utils/compat/numpycompat.py:7]: I don't think the changes in `numpycompat` are used in this PR, so it should be separate.

**Review by mhvk — CHANGES_REQUESTED** [review]
I like splitting up the module, but am less sure about also moving tests (partially since that means my memory of where test files lives will go haywire, while for representations itself, I'll tab-complete my way easily). I also think the tests themselves could do with a bit of reorganization, so I'd suggest separating the two steps. (An additional advantage is that then the test files will help ensure all `from astropy.coordinates.reoresentation import ???` continue to work.)

I would also *not* split the various spherical representations - they are so clearly related that I think it is easier if they stay together (e.g., to keep the various quick transformations in one place), i.e., I would combine back `radial`, `spherical`, `base_sphere`, `physicssphere` and `unitsphere`, leaving as new modules `base`, `cartesian`, `cylindrical`, and `spherical`.

File: astropy/coordinates/representation/__init__.py
Comment (nstarman) [reply astropy/coordinates/representation/__init__.py:17]: SGTM.

File: astropy/coordinates/representation/cylindrical.py
Comment (nstarman) [reply astropy/coordinates/representation/cylindrical.py:17]: I include an empty ``__all__`` to:

1. To show that this module is private. Imports should happen from ``astropy.coordinates`` or ``astropy.coordinates.representation`` (this is included for backwards compatibility).
2. Prevent bad dumps in a namespace via ``from ... import *``.

Python thinks ``__all__`` should be a ``list[str]`` but of course does not enforce it. Empty lists need to be typed to demonstrate compliance with Python expectations.

File: astropy/coordinates/representation/tests/test_representation.py
Comment (nstarman) [reply astropy/coordinates/representation/tests/test_representation.py:2]: Moving back....

Comment (nstarman) [reply astropy/coordinates/representation/tests/test_representation.py:2]: ✅

**Comment by pllim** [comment]
Re: doctestplus -- no idea, hard to see with `+3,796 −3,523`

File: astropy/table/tests/test_mixin.py
Comment (pllim) [inline astropy/table/tests/test_mixin.py:15]: Isn't this technically breaking API?

File: astropy/coordinates/representation/__init__.py
Comment (nstarman) [reply astropy/coordinates/representation/__init__.py:17]: ✅

File: astropy/coordinates/representation/cartesian.py
Comment (nstarman) [reply astropy/coordinates/representation/cartesian.py:19]: Getting rid of them everywhere. It's from my python file template. Nice and organized.

Comment (nstarman) [reply astropy/coordinates/representation/cartesian.py:19]: ✅

File: astropy/table/tests/test_mixin.py
Comment (nstarman) [reply astropy/table/tests/test_mixin.py:15]: I didn't think they were.

File: astropy/coordinates/representation/physicsphere.py
Comment (nstarman) [reply astropy/coordinates/representation/physicsphere.py:20]: Getting rid of ``# CODE``. 
I like the ``__all__``. It's makes our intent very explicit.

File: astropy/table/tests/test_mixin.py
Comment (pllim) [reply astropy/table/tests/test_mixin.py]: When different subpackages import from it, it is kinda semi-public. There is a chance downstream package test might be using it too but I am not sure.

**Comment by nstarman** [comment]
Interesting. De-serialization is an issue since the qualitative name has changed.

**Comment by pllim** [comment]
I just reran the last CI on `main` to see if the failure is caused by your changes or something else: https://github.com/astropy/astropy/actions/runs/3069379004

**Comment by pllim** [comment]
`main` still is green, so it is definitely you... 😅

File: astropy/coordinates/representation/cylindrical.py
Comment (mhvk) [reply astropy/coordinates/representation/cylindrical.py:14]: OK on `__all__` - can intent perhaps be even more explicit by adding a comment? Maybe something like
```
# Outside users should import from astropy.coordinates.representations.
```

File: astropy/table/tests/test_mixin.py
Comment (mhvk) [reply astropy/table/tests/test_mixin.py]: I think this goes away if the tests stay in place; maybe another reason to think about splitting up `representations` and moving the tests separately.

Comment (nstarman) [reply astropy/table/tests/test_mixin.py]: As in 2 separate PRs?

**Comment by taldcroft** [comment]
About the test failure
```
        if cls_full_name not in __construct_mixin_classes:
>           raise ValueError(f'unsupported class for construct {cls_full_name}')
E           ValueError: unsupported class for construct astropy.coordinates.representation.cartesian.CartesianRepresentation
```
This is assuming the class full name changed. There is a hard-coded lists of classes that the YAML constructor is allowed to construct, which is there for security reasons. It's hard-coded as strings so table doesn't have to import the rest of astropy always. @mhvk have talked about this before and never come up with a better way. 

It's probably OK to just changed that hardcoded list, but doing stuff like this always makes you wonder if you will be breaking anything outside of astropy.

File: astropy/table/tests/test_mixin.py
Comment (mhvk) [reply astropy/table/tests/test_mixin.py]: Yes, sorry that I wasn't clear -- I think the splitting of the module is easy as long as the tests still run. The moving of the tests I'm not sure is best, but if we do it, might make sense to reorganize a bit.

**Comment by mhvk** [comment]
Oops, I guess that also means that we have to continue to support both `astropy.coordinates.representation.CartesianRepresentation` as well as the same but with `cartesian` inserted. Unless we start to muck with the `__module__` in this PR, but I've never liked if things do not point to where they are defined. Another option might be for `__construct_mixin_classes` to allow wildcards.

**Comment by pllim** [comment]
Maaaybe we need to hold off on this until we unpin Sphinx. The coordinate imports galore already broke compat with Sphinx. I don't want this to break it more.

**Comment by nstarman** [comment]
I opened https://github.com/astropy/astropy/issues/13689 to discuss improvements to ``__construct_mixin_classes``.

File: astropy/coordinates/representation/cylindrical.py
Comment (nstarman) [reply astropy/coordinates/representation/cylindrical.py]: ✅

File: astropy/table/tests/test_mixin.py
Comment (nstarman) [reply astropy/table/tests/test_mixin.py:15]: ✅

**Comment by nstarman** [comment]
@mhvk @pllim @taldcroft @eerovaher

Ok. One optional item left on the checklist: change folder name to ``representations``. This is a breaking change, but brings the folder more in line with the contents and other folders, like ``builtin_frames``.

**Comment by mhvk** [comment]
My sense would be to leave it as `representation` - I agree the plural is more logical but we have not been very consistent (`quantity.py`,  `sky_coordinate.py`, though in both cases there is really only one class inside) and one of the nice things of the present approach is that zero work is required outside of `representation` itself.

**Comment by mhvk** [comment]
p.s. And one can think of the module being *about* `representation` rather than containing multiple `representations`.

**Review by pllim — CHANGES_REQUESTED** [review]
Comparing this with built-in frame made me nervous. I would like to not merge this until https://github.com/astropy/astropy/issues/11725 is resolved and we can make sure this won't reintroduce that incompability.

File: astropy/coordinates/representation/cylindrical.py
Comment (astrofrog) [inline astropy/coordinates/representation/cylindrical.py:1]: I think all files should include the license line at the top, so make sure you add these to new files here. I wonder if we can add this check to pre-commit?

**Review by astrofrog — COMMENTED** [review]
I'm not really qualified to review this, but just a small comment below.

File: astropy/coordinates/representation/cylindrical.py
Comment (nstarman) [reply astropy/coordinates/representation/cylindrical.py:1]: We could make a pre-commit hook that can autocorrect this.

Comment (eerovaher) [reply astropy/coordinates/representation/cylindrical.py:1]: https://pypi.org/project/pre-commit-license-headers/

Comment (nstarman) [reply astropy/coordinates/representation/cylindrical.py:1]: Closing the because we have #13786

File: astropy/coordinates/builtin_frames/galactocentric.py
Comment (mhvk) [inline astropy/coordinates/builtin_frames/galactocentric.py:360]: Hmm, none of these changes should be necessary (after all, we can use `~astropy.units.Quantity`), but perhaps the sphinx documentation settings are not correct? Or would just `~astropy.coordinates.CartesianDifferential` work?

Comment (nstarman) [reply astropy/coordinates/builtin_frames/galactocentric.py:360]: A good point. I'll try a TMP push with it reverted.

**Comment by eteq** [comment]
I  like the idea of making this change work if we can but it's 5.2 freeze day and I'm not clear if you're still making changes, @nstarman ? (I can try to review it this evening if so, although I do have the misgivings that some of the issues above might mean breaking external compatibility.) 

(I know we're still pinning sphinx but... :shrug: maybe we need to live dangrously?)

**Comment by nstarman** [comment]
I think this is best done for v5.3, given the compatibility concerns.

**Comment by mhvk** [comment]
I think this is definitely a good thing to do, but there is no particular reason to do it before a release (except perhaps not to have to redo the work after the black formatting...). My sense would be to return to this a bit after the release.

File: astropy/coordinates/builtin_frames/galactocentric.py
Comment (nstarman) [reply astropy/coordinates/builtin_frames/galactocentric.py]: I've had success with ``astropy.coordinates...CartesianDifferential``

**Comment by nstarman** [comment]
v5.2 is out, so this PR is ready to go in!

**Comment by pllim** [comment]
@mhvk requested changes in the review status, so maybe he wants to re-review?

And just to make sure, you did rebase recently enough to pick up the Sphinx unpinning, right?

**Comment by nstarman** [comment]
I believe so. That was some time last week, right?

File: astropy/coordinates/representation/base.py
Comment (eerovaher) [inline astropy/coordinates/representation/base.py:730]: Why is this split into two strings? It's a single string in `main`: https://github.com/astropy/astropy/blob/2fab3d7a3e51cda1fdcf61fdd36ab5087f894541/astropy/coordinates/representation.py#L745-L747

File: astropy/coordinates/representation/base.py
Comment (eerovaher) [inline astropy/coordinates/representation/base.py:719]: On `main` this error message is written in two lines, both within the 88 character line length limit, so there's no reason to split this into three lines here.
https://github.com/astropy/astropy/blob/2fab3d7a3e51cda1fdcf61fdd36ab5087f894541/astropy/coordinates/representation.py#L733-L736

**Review by eerovaher — COMMENTED** [review]
Splitting up the representation classes into separate files is a good idea, but such a pull request that is moving lots of code around should try to avoid making any changes that are not necessary, at least not in the same commit that moves the code. @nstarman, it looks like you have some automatic tool (at least I hope it's an automatic tool) that edits comments and multi-line strings so that they fit into an 80 character line length limit. This does not affect the working of the code, but it is creating a lot of clutter in the diffs. Furthermore, the actual line length limit in `astropy` is now 88 characters. I've only marked two cases where the difference of enforced line lengths has led to needless editing of multi-line strings, but I expect there are more.

**Comment by nstarman** [comment]
@eerovaher, I did the latter and then went and compressed error message strings into the minimum number of lines.

File: astropy/coordinates/representation/base.py
Comment (nstarman) [reply astropy/coordinates/representation/base.py]: Fixed!

File: astropy/coordinates/representation/base.py
Comment (nstarman) [reply astropy/coordinates/representation/base.py]: Fixed!

File: astropy/coordinates/representation/__init__.py
Comment (mhvk) [reply astropy/coordinates/representation/__init__.py:34]: I'm with @eerovaher -- as few changes here as possible. So, let's keep this the same as the one originally in `representations.py`.

File: astropy/coordinates/representation/base.py
Comment (mhvk) [inline astropy/coordinates/representation/base.py:17]: I disagree with this still. Let's put things in here that get transferred up (like is done in modules in `builtin_frames`). If you feel strongly otherwise, I would suggest for now not having any `__all__` (like in `matrix_utilities`, which is also not exposed).  Though to me it is helpful to the code reader -- the ones for whom this matters most -- that it lists the classes, etc., that this file actually defines for use elsewhere.

File: astropy/coordinates/representation/base.py
Comment (mhvk) [inline astropy/coordinates/representation/base.py:34]: Oddly, changed line.

File: astropy/coordinates/representation/cartesian.py
Comment (mhvk) [inline astropy/coordinates/representation/cartesian.py:10]: As above

File: astropy/coordinates/representation/cylindrical.py
Comment (mhvk) [reply astropy/coordinates/representation/cylindrical.py:1]: So, why is the license file still not added automatically then?

File: astropy/coordinates/representation/base.py
Comment (mhvk) [inline astropy/coordinates/representation/base.py:1]: I think there is no need for this if we remove `__all__`.

File: astropy/coordinates/representation/spherical.py
Comment (mhvk) [inline astropy/coordinates/representation/spherical.py:18]: As above.

File: docs/coordinates/skycoord.rst
Comment (mhvk) [inline docs/coordinates/skycoord.rst:764]: I think this is better changed to `~astropy.coordinates.BaseRepresentation`, since that is where we want users to import from.

**Review by mhvk — COMMENTED** [review]
Looks mostly OK, but I'm still unhappy with the empty `__all__` -- readability counts, and this is done *not* for people reading the code, for whom it would be useful to know what classes are actually the ones being defined that get used elsewhere. Plus, that's how its done elsewhere in astropy.

File: astropy/coordinates/representation/base.py
Comment (eerovaher) [inline astropy/coordinates/representation/base.py:1040]: One f-string is good enough for `main`, it should be good enough here as well.
https://github.com/astropy/astropy/blob/87963074a50b14626fbd825536f384a6e96835af/astropy/coordinates/representation.py#L1070

**Review by eerovaher — COMMENTED** [review]
I've highlighted a couple of needless changes to strings that are still present, but I have not made a thorough search even in the one file I looked at, so I expect there to be many more cases of this.

@nstarman, the tool that you are using now seems to respect the 88 character line length limit, but it is being very aggressive about it, i.e. it adds words from the beginning of the next line to the end of the current line if they fit. This is being done in multi-line strings, comments and docstrings.

I supported adopting `black` because I was annoyed with the number of commits in this repository that introduced formatting problems that had to be fixed in a separate commit. Such formatting fixes add clutter to the commit log and make it more difficult to track actual functional changes. `black` together with `pre-commit` can prevent such formatting-only commits from ever happening (although not everyone is making use of these tools). The tool you are using is making cosmetic changes to a codebase that `black` is happy with. This is adding clutter to the commit history, which I cannot approve. I recommend you turn off any code-formatting features VSCode might have and just let `black` do its job.

**Comment by pllim** [comment]
Wow `+3,728 −3,658`. I think this is too late to squeeze into v5.3 only 3 days before feature freeze. I am re-milestoning this.

I would recommend you wrap this up early in the v6.0 dev cycle. Or maybe breaking it into a series of smaller PRs.

**Comment by mhvk** [comment]
See #14792 for a re-implementation of this - I still think this is an excellent idea but this PR was getting behind `main` and there were still a few additions that were a little controversial. I tried to keep it as much of a copy & paste effort as possible.

**Pull request closed without merging**
